tpu-client/src/tpu_connection_cache.rs

```diff
@@ -4,9 +4,54 @@ use {
         nonblocking::tpu_connection::TpuConnection as NonblockingTpuConnection,
         tpu_connection::TpuConnection as BlockingTpuConnection,
     },
+    rand::{thread_rng, Rng},
     std::{net::SocketAddr, sync::Arc},
+    thiserror::Error,
 };
 
+#[derive(Error, Debug)]
+pub enum ConnectionPoolError {
+    #[error("connection index is out of range of the pool")]
+    IndexOutOfRange,
+}
+
+pub trait ConnectionPool {
+    type PoolTpuConnection: BaseTpuConnection;
+    type TpuConfig: Default;
+    const PORT_OFFSET: u16 = 0;
+
+    /// Create a new connection pool based on protocol-specific configuration
+    fn new_with_connection(config: &Self::TpuConfig, addr: &SocketAddr) -> Self;
+
+    /// Add a connection to the pool
+    fn add_connection(&mut self, config: &Self::TpuConfig, addr: &SocketAddr);
+
+    /// Get the number of current connections in the pool
+    fn num_connections(&self) -> usize;
+
+    /// Get a connection based on its index in the pool, without checking if the
+    fn get(&self, index: usize) -> Result<Arc<Self::PoolTpuConnection>, ConnectionPoolError>;
+
+    /// Get a connection from the pool. It must have at least one connection in the pool.
+    /// This randomly picks a connection in the pool.
+    fn borrow_connection(&self) -> Arc<Self::PoolTpuConnection> {
+        let mut rng = thread_rng();
+        let n = rng.gen_range(0, self.num_connections());
+        self.get(n).expect("index is within num_connections")
```

Comment:
This retains current ConnectionCache pool functionality, except using `get().expect()` instead of a direct index into the pool:
https://github.com/solana-labs/solana/blob/f38ed1c26638f6afcab882ce393db99629467e91/tpu-client/src/connection_cache.rs#L257-L258

I added `get()` and use it here to prep for fallible methods, but would prefer to make deliberate functionality changes separately to help prevent accidental functionality changes in this PR.

That said, the pool IndexMap is behind a RwLock, so as long as the cache methods aren't moving the individual `ConnectionPool`s into threads after unlock, I don't think the count can change here.

Reply:
cool. cool. if there are plans for future fallibility, fine by me!